relay-server/src/actors/project_cache.rs

```diff
@@ -217,11 +180,37 @@ impl Message for FetchOptionalProjectState {
     type Result = Option<Arc<ProjectState>>;
 }
 
-impl Handler<FetchProjectState> for ProjectCache {
-    type Result = Response<ProjectStateResponse, ()>;
+/// Fetches a project state from one of the available sources.
+///
+/// The project state is resolved in the following precedence:
+///
+///  1. Local file system
+///  2. Redis cache (processing mode only)
+///  3. Upstream (managed and processing mode only)
+///
+/// Requests to the upstream are performed via `UpstreamProjectSource`, which internally batches
+/// individual requests.
+#[derive(Clone)]
+pub struct UpdateProjectState {
+    /// The public key to fetch the project by.
+    pub public_key: ProjectKey,
+
+    /// If true, all caches should be skipped and a fresh state should be computed.
+    pub no_cache: bool,
+}
+
+impl Message for UpdateProjectState {
+    type Result = ();
+}
+
+impl Handler<UpdateProjectState> for ProjectCache {
+    type Result = ();
 
-    fn handle(&mut self, message: FetchProjectState, _context: &mut Self::Context) -> Self::Result {
-        let public_key = message.public_key;
+    fn handle(&mut self, message: UpdateProjectState, context: &mut Self::Context) -> Self::Result {
+        let UpdateProjectState {
+            public_key,
+            no_cache,
+        } = message;
 
         if let Some(mut entry) = self.projects.get_mut(&public_key) {
```

Comment:
Ccall `self.get_project` (get_or_create) here to ensure the project exists.

Reply:
ok